### ISSUE TITLE
No addDir/add event when a directory is renamed on OSX

### ISSUE THREAD

**Issue by nono** [paulmillr/chokidar#467]
Hello,

it looks like chokidar emits no `addDir` event when a directory is renamed on OSX (and no `add` for a file inside it).

The JS script:

```js
#!/usr/bin/env node

const chokidar = require('chokidar')
const path = require('path')

const log = console.log.bind(console)
const workspace = path.resolve('./workspace')

log('workspace is', workspace)

const watcher = chokidar.watch('.', {
  cwd: workspace,
  ignored: /[\/\\]\.cozy-desktop/,
  followSymlinks: false,
  alwaysStat: true,
  atomic: true,
  awaitWriteFinish: { pollInterval: 200, binaryInterval: 1000 }
})

// Show events
watcher
  .on('add', (path, stats) => { log('add', path) })
  .on('addDir', (path, stats) => { log('addDir', path) })
  .on('change', (path, stats) => { log('change', path) })
  .on('unlink', (path) => { log('unlink', path) })
  .on('unlinkDir', (path) => { log('unlinkDir', path) })
  .on('error', (err) => { log('error', err) })
  .on('ready', () => { log('ready') })
  .on('raw', (event, path, details) => { log('raw', event, path, details) })

// After 15s, print the watchers
setTimeout( () => { log('watchers:', watcher.getWatched()) }, 15000)

var i = 0
setInterval( () => { log(++i + 's') }, 1000)
```

It is launched by this script:

```bash
#!/bin/bash

set -eux

rm -rf workspace
mkdir -p workspace/.cozy-desktop

node test-chokidar.js &

cd workspace

sleep 1
mkdir aaa
date > aaa/bbb

sleep 1
mkdir .cozy-desktop/zzz
date > .cozy-desktop/zzz/xxx
date > .cozy-desktop/yyy

sleep 4
mv aaa ccc

sleep 10
ls -alR
cd ..
kill %1
```

The output:

```
+ rm -rf workspace
+ mkdir -p workspace/.cozy-desktop
+ cd workspace
+ sleep 1
+ node test-chokidar.js
workspace is /Users/cozycloud/code/tests/workspace
addDir 
ready
+ mkdir aaa
+ date
MacBook-Pro-de-Cozy:tests cozycloud$ ./run.sh 
+ rm -rf workspace
+ mkdir -p workspace/.cozy-desktop
+ cd workspace
+ sleep 1
+ node test-chokidar.js
workspace is /Users/cozycloud/code/tests/workspace
addDir 
ready
+ mkdir aaa
+ date
+ sleep 1
1s
raw unknown /Users/cozycloud/code/tests/workspace/aaa { path: '/Users/cozycloud/code/tests/workspace/aaa',
  event: 'unknown',
  type: 'directory',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 131328 }
raw modified /Users/cozycloud/code/tests/workspace/aaa/bbb { path: '/Users/cozycloud/code/tests/workspace/aaa/bbb',
  event: 'modified',
  type: 'file',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 69888 }
+ mkdir .cozy-desktop/zzz
+ date
+ date
+ sleep 4
2s
raw deleted /Users/cozycloud/code/tests/workspace/.cozy-desktop/zzz { path: '/Users/cozycloud/code/tests/workspace/.cozy-desktop/zzz',
  event: 'deleted',
  type: 'directory',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 131840 }
raw deleted /Users/cozycloud/code/tests/workspace/.cozy-desktop/zzz/xxx { path: '/Users/cozycloud/code/tests/workspace/.cozy-desktop/zzz/xxx',
  event: 'deleted',
  type: 'file',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 70400 }
raw deleted /Users/cozycloud/code/tests/workspace/.cozy-desktop/yyy { path: '/Users/cozycloud/code/tests/workspace/.cozy-desktop/yyy',
  event: 'deleted',
  type: 'file',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 70400 }
3s
add aaa/bbb
4s
5s
+ mv aaa ccc
+ sleep 10
6s
raw moved /Users/cozycloud/code/tests/workspace/aaa { path: '/Users/cozycloud/code/tests/workspace/aaa',
  event: 'moved',
  type: 'directory',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 133376 }
raw deleted /Users/cozycloud/code/tests/workspace/ccc { path: '/Users/cozycloud/code/tests/workspace/ccc',
  event: 'deleted',
  type: 'directory',
  changes: { inode: false, finder: false, access: false, xattrs: false },
  flags: 133632 }
unlinkDir aaa
unlink aaa/bbb
7s
8s
9s
10s
11s
12s
13s
14s
watchers: { '..': [ 'workspace' ],
  '.': [ 'ccc' ],
  '.cozy-desktop/zzz': [],
  '.cozy-desktop': [],
  ccc: [] }
15s
+ ls -alR
total 0
drwxr-xr-x  4 cozycloud  staff  136 22 avr 11:12 .
drwxr-xr-x  7 cozycloud  staff  238 22 avr 11:11 ..
drwxr-xr-x  4 cozycloud  staff  136 22 avr 11:12 .cozy-desktop
drwxr-xr-x  3 cozycloud  staff  102 22 avr 11:12 ccc

./.cozy-desktop:
total 8
drwxr-xr-x  4 cozycloud  staff  136 22 avr 11:12 .
drwxr-xr-x  4 cozycloud  staff  136 22 avr 11:12 ..
-rw-r--r--  1 cozycloud  staff   30 22 avr 11:12 yyy
drwxr-xr-x  3 cozycloud  staff  102 22 avr 11:12 zzz

./.cozy-desktop/zzz:
total 8
drwxr-xr-x  3 cozycloud  staff  102 22 avr 11:12 .
drwxr-xr-x  4 cozycloud  staff  136 22 avr 11:12 ..
-rw-r--r--  1 cozycloud  staff   30 22 avr 11:12 xxx

./ccc:
total 8
drwxr-xr-x  3 cozycloud  staff  102 22 avr 11:12 .
drwxr-xr-x  4 cozycloud  staff  136 22 avr 11:12 ..
-rw-r--r--  1 cozycloud  staff   30 22 avr 11:12 bbb
+ cd ..
+ kill %1
```

By the way, is it normal to have watchers on `.cozy-desktop` and `.cozy-desktop/zzz` with `ignored: /[\/\\]\.cozy-desktop/`?

**Comment by nono** [comment]
Chokidar was installed from the `master` branch.

**Comment by nono** [comment]
With https://github.com/paulmillr/chokidar/pull/468/commits/319da98025369709989e134b148e6b764b3098fa, I have the missing `addDir aaa` before renaming the folder, and `addDir ccc` after renaming it. But the `add ccc/bbb` is still missing.

If I create a sub-directory `ddd` inside `aaa`, I have no events `addDir ccc/ddd` after the move.

**Comment by nono** [comment]
#475 should fix the missing `addDir`. But, if the directory had some files inside it, we have `unlink` events for them but no `add`.

**Comment by nono** [comment]
And #490 should help for the missing `add` events.

**Comment by es128** [comment]
Resolved by #490. Thanks @nono.
